src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyNode.cs

```diff
@@ -193,47 +167,106 @@ public virtual string Alias
 
         public int Priority { get; protected set; }
 
-        public ProjectTreeFlags Flags { get; set; }
+        public ProjectTreeFlags Flags { get; internal set; }
 
         public IImmutableDictionary<string, string> Properties { get; protected set; }
 
-        private HashSet<IDependencyNode> _children = new HashSet<IDependencyNode>();
-        public HashSet<IDependencyNode> Children
+        private ImmutableHashSet<IDependencyNode> _children = ImmutableHashSet<IDependencyNode>.Empty;
+        public ImmutableHashSet<IDependencyNode> Children
         {
             get
             {
-                return _children;
+                lock (_children)
```

Comment:
Should have a proper sync here rather than locking the set itself - especially since AddChild locks the set it replaces